README.md
primecount
==========
[![Build Status](https://travis-ci.org/kimwalisch/primecount.svg)](https://travis-ci.org/kimwalisch/primecount)
[![Build Status](https://ci.appveyor.com/api/projects/status/github/kimwalisch/primecount?branch=master&svg=true)](https://ci.appveyor.com/project/kimwalisch/primecount)
[![GitHub license](https://img.shields.io/badge/license-BSD%202-blue.svg)](https://github.com/kimwalisch/primecount/blob/master/COPYING)

primecount is a command-line program and C++ library that counts the
primes below an integer x&nbsp;≤&nbsp;10<sup>31</sup> using **highly
optimized** implementations of the
[prime counting function](http://en.wikipedia.org/wiki/Prime-counting_function)
(combinatorial methods). primecount includes implementations of the algorithms
of Legendre, Meissel, Lehmer, Lagarias-Miller-Odlyzko and Deleglise-Rivat all
of which have been parallelized using
[OpenMP](http://en.wikipedia.org/wiki/OpenMP). The Deleglise-Rivat
implementation has 
[recently been distributed](https://github.com/kimwalisch/primecount/blob/master/doc/primecount-MPI.md#primecount-mpi)
using MPI.

primecount contains the **first ever** parallel open source
implementation of the Deleglise-Rivat algorithm and it features a
[novel load balancer](https://github.com/kimwalisch/primecount/blob/master/src/S2LoadBalancer.cpp)
which scales up to hundreds of CPU cores. primecount has already been
used to compute several world records e.g.
[pi(10<sup>27</sup>)](http://www.mersenneforum.org/showthread.php?t=20473) and
[nth_prime(10<sup>24</sup>)](https://oeis.org/A006988), more will hopefully follow!

Binaries
--------
Below are the latest precompiled binaries for Windows 64-bit, Linux and OS X.
These binaries are statically linked and require a CPU which supports the POPCNT
instruction (2010 or later).

* [primecount-3.5-win64.zip](https://dl.bintray.com/kimwalisch/primecount/primecount-3.5-win64.zip), 391K
* [primecount-3.5-linux-x64.tar.gz](https://dl.bintray.com/kimwalisch/primecount/primecount-3.5-linux-x64.tar.gz), 902K
* [primecount-3.5-macOS-x64.tar.gz](https://dl.bintray.com/kimwalisch/primecount/primecount-3.5-macOS-x64.tar.gz), 673K
* Binaries with backup functionality are available [here](https://github.com/kimwalisch/primecount/tree/backup#primecount-backup)

Usage examples
--------------
Open a terminal and run primecount using e.g.:
```sh
# Count the primes below 10^14
./primecount 1e14

# Print progress and status information during computation
./primecount 1e20 --status

# Count primes using Meissel's algorithm
./primecount 2**32 --meissel

# Find the 10^14th prime using 4 threads
./primecount 1e14 --nthprime --threads=4 --time
```

Command-line options
--------------------
```
Usage: primecount x [OPTION]...
Count the primes below x <= 10^31 using fast implementations of the
combinatorial prime counting function.

Options:

  -d,    --deleglise_rivat  Count primes using Deleglise-Rivat algorithm
         --legendre         Count primes using Legendre's formula
         --lehmer           Count primes using Lehmer's formula
  -l,    --lmo              Count primes using Lagarias-Miller-Odlyzko
  -m,    --meissel          Count primes using Meissel's formula
         --Li               Approximate pi(x) using the logarithmic integral
         --Li_inverse       Approximate the nth prime using Li^-1(x)
  -n,    --nthprime         Calculate the nth prime
  -p,    --primesieve       Count primes using the sieve of Eratosthenes
  -s[N], --status[=N]       Show computation progress 1%, 2%, 3%, ...
                            [N] digits after decimal point e.g. N=1, 99.9%
         --test             Run various correctness tests and exit
         --time             Print the time elapsed in seconds
  -t<N>, --threads=<N>      Set the number of threads, 1 <= N <= CPU cores
  -v,    --version          Print version and license information
  -h,    --help             Print this help menu

Advanced Deleglise-Rivat options:

  -a<N>, --alpha=<N>        Tuning factor, 1 <= alpha <= x^(1/6)
         --P2               Only compute the 2nd partial sieve function
         --S1               Only compute the ordinary leaves
         --S2_trivial       Only compute the trivial special leaves
         --S2_easy          Only compute the easy special leaves
         --S2_hard          Only compute the hard special leaves
```

Algorithms
----------
<table>
  <tr>
    <td>Legendre's Formula</td>
    <td><img src="http://kimwalisch.github.io/primecount/formulas/pi_legendre.svg" height="20" align="absmiddle"/></td>
  </tr>
  <tr>
    <td>Meissel's Formula</td>
    <td><img src="http://kimwalisch.github.io/primecount/formulas/pi_meissel.svg" height="20" align="absmiddle"/></td>
  </tr>
  <tr>
    <td>Lehmer's Formula</td>
    <td><img src="http://kimwalisch.github.io/primecount/formulas/pi_lehmer.svg" height="20" align="absmiddle"/></td>
  </tr>
  <tr>
    <td>LMO Formula</td>
    <td><img src="http://kimwalisch.github.io/primecount/formulas/pi_lmo.svg" height="20" align="absmiddle"/></td>
  </tr>
</table>

<p>Up until the early 19th century the most efficient known method for counting
primes was the sieve of Eratosthenes which has a running time of
<img src="http://kimwalisch.github.io/primecount/formulas/Oxloglogx.svg" height="20" align="absmiddle"/>
operations. The first improvement to this bound was Legendre's formula (1830)
which uses the inclusion-exclusion principle to calculate the number of primes
below x without enumerating the individual primes. Legendre's formula has a
running time of
<img src="http://kimwalisch.github.io/primecount/formulas/Ox.svg" height="20" align="absmiddle"/>
operations and uses
<img src="http://kimwalisch.github.io/primecount/formulas/Osqrtx.svg" height="20" align="absmiddle"/>
space. In 1870 E. D. F. Meissel improved Legendre's formula by setting
<img src="http://kimwalisch.github.io/primecount/formulas/apisqrt3x.svg" height="20" align="absmiddle"/>
and by adding the correction term
<img src="http://kimwalisch.github.io/primecount/formulas/P2xa.svg" height="20" align="absmiddle"/>.
Meissel's formula has a running time of
<img src="http://kimwalisch.github.io/primecount/formulas/Omeissel.svg" height="20" align="absmiddle"/>
operations and uses
<img src="http://kimwalisch.github.io/primecount/formulas/Osqrtxlogx.svg" height="20" align="absmiddle"/>
space. In 1959 D. H. Lehmer extended Meissel's formula and slightly improved the running time to
<img src="http://kimwalisch.github.io/primecount/formulas/Olehmer.svg" height="20" align="absmiddle"/>
operations and
<img src="http://kimwalisch.github.io/primecount/formulas/Osqrtxlogx.svg" height="20" align="absmiddle"/>
space. In 1985 J. C. Lagarias, V. S. Miller and A. M. Odlyzko published a new
algorithm based on Meissel's formula which has a lower runtime complexity of
<img src="http://kimwalisch.github.io/primecount/formulas/Oroot23xlogx.svg" height="20" align="absmiddle"/>
operations and which uses only
<img src="http://kimwalisch.github.io/primecount/formulas/Osqrt3xlog2x.svg" height="20" align="absmiddle"/>
space.</p>
<p>primecount's Legendre, Meissel and Lehmer implementations are based on
Hans Riesel's book <a href="https://github.com/kimwalisch/primecount#references">[5]</a>,
its Lagarias-Miller-Odlyzko and Deleglise-Rivat implementations are
based on Tomás Oliveira's paper
<a href="https://github.com/kimwalisch/primecount#references">[8]</a>.</p>

Fast nth prime calculation
--------------------------
The most efficient known method for calculating the nth prime is a combination
of the prime counting function and a prime sieve. The idea is to closely
approximate the nth prime using e.g. the inverse logarithmic integral
<img src="http://kimwalisch.github.io/primecount/formulas/Li-1n.svg" height="20" align="absmiddle"/>
and then count the primes up to this guess using the prime counting function.
Once this is done one starts sieving (e.g. using the segmented sieve of
Eratosthenes) from there on until one finds the actual nth prime. The author
has implemented ```primecount::nth_prime(n)``` this way, it finds the nth
prime in
<img src="http://kimwalisch.github.io/primecount/formulas/Oroot23xlog2x.svg" height="20" align="absmiddle"/>
operations using
<img src="http://kimwalisch.github.io/primecount/formulas/Opisqrtx.svg" height="20" align="absmiddle"/>
space.

Benchmarks
----------
<table>
  <tr align="center">
    <td><b>x</b></td>
    <td><b>Prime Count</b></td>
    <td><b>Legendre</b></td>
    <td><b>Meissel</b></td>
    <td><b>Lagarias<br/>Miller<br/>Odlyzko</b></td>
    <td><b>Deleglise<br/>Rivat</b></td>
  </tr>
  <tr align="right">
    <td>10<sup>10</sup></td>
    <td>455,052,511</td>
    <td>0.02s</td>
    <td>0.01s</td>
    <td>0.01s</td>
    <td>0.01s</td>
  </tr>
  </tr>
  <tr align="right">
    <td>10<sup>11</sup></td>
    <td>4,118,054,813</td>
    <td>0.04s</td>
    <td>0.05s</td>
    <td>0.02s</td>
    <td>0.02s</td>
  </tr>
  </tr>
  <tr align="right">
    <td>10<sup>12</sup></td>
    <td>37,607,912,018</td>
    <td>0.15s</td>
    <td>0.13s</td>
    <td>0.03s</td>
    <td>0.02s</td>
  </tr>
  </tr>
  <tr align="right">
    <td>10<sup>13</sup></td>
    <td>346,065,536,839</td>
    <td>0.70s</td>
    <td>0.46s</td>
    <td>0.10s</td>
    <td>0.07s</td>
  </tr>
  </tr>
  <tr align="right">
    <td>10<sup>14</sup></td>
    <td>3,204,941,750,802</td>
    <td>4.01s</td>
    <td>1.96s</td>
    <td>0.43s</td>
    <td>0.22s</td>
  </tr>
  <tr align="right">
    <td>10<sup>15</sup></td>
    <td>29,844,570,422,669</td>
    <td>27.75s</td>
    <td>12.08s</td>
    <td>1.65s</td>
    <td>0.77s</td>
  </tr>
  <tr align="right">
    <td>10<sup>16</sup></td>
    <td>279,238,341,033,925</td>
    <td>232.30s</td>
    <td>92.09s</td>
    <td>7.31s</td>
    <td>2.83s</td>
  </tr>
  <tr align="right">
    <td>10<sup>17</sup></td>
    <td>2,623,557,157,654,233</td>
    <td>1,836.73s</td>
    <td>731.35s</td>
    <td>33.66s</td>
    <td>11.20s</td>
  </tr>
  <tr align="right">
  <td>10<sup>18</sup></td>
  <td>24,739,954,287,740,860</td>
    <td>14,949.16s</td>
    <td>6,631.73s</td>
    <td>157.15s</td>
    <td>47.07s</td>
  </tr>
  <tr align="right">
  <td>10<sup>19</sup></td>
  <td>234,057,667,276,344,607</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>225.08s</td>
  </tr>
  <tr align="right">
  <td>10<sup>20</sup></td>
  <td>2,220,819,602,560,918,840</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>1,029.41s</td>
  </tr>
  <tr align="right">
  <td>10<sup>21</sup></td>
  <td>21,127,269,486,018,731,928</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>4,867.45s</td>
  </tr>
  <tr align="right">
  <td>10<sup>22</sup></td>
  <td>201,467,286,689,315,906,290</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>22,361.90s</td>
  </tr>
</table>

The benchmarks above were run on an Intel Core i7-6700 CPU (4 x 3.4 GHz) from
2015 using a Linux x64 operating system and primecount was compiled using
GCC 5.2.

Build instructions (Unix-like OSes)
-----------------------------------
You need to have installed a C++ compiler and GNU make to build primecount.

Download
[primecount-3.5.tar.gz](https://dl.bintray.com/kimwalisch/primecount/primecount-3.5.tar.gz)
and build it using:

```sh
./build.sh
make check
sudo make install
```

To build primecount using
[MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface)
support for distributing computations onto cluster nodes use:
```sh
./build.sh --enable-mpi
```

[primecount-MPI.md](doc/primecount-MPI.md) contains more information.

Build instructions (Microsoft Visual C++)
-----------------------------------------
Open a Visual Studio Command Prompt and run:
```sh
nmake -f Makefile.msvc
```

C++ API
-------
Below are the main functions declared in
[primecount.hpp](https://github.com/kimwalisch/primecount/blob/master/include/primecount.hpp).
All functions are multi-threaded by default.

```C++
/// Count the primes <= x
int64_t primecount::pi(int64_t x);

/// 128-bit prime counting function.
/// @param expr  Integer arithmetic expression e.g. "1000", "10^22"
/// @pre   expr  <= 10^31 on 64-bit systems
///        expr    < 2^63 on 32-bit systems
std::string primecount::pi(const std::string& expr);

/// Find the nth prime
int64_t primecount::nth_prime(int64_t n);
```

C++ library usage
-----------------
Below is an example program that counts the primes below 1000.

```C++
#include <primecount.hpp>
#include <iostream>

int main()
{
    int64_t prime_count = primecount::pi(1000);
    std::cout << "primes below 1000 = " << prime_count << std::endl;
  
    return 0;
}
```

On Unix-like OSes compile using:
```sh
c++ -O2 primes.cpp -lprimecount -lprimesieve
```

References
----------
1. A. M. Legendre, Théorie des nombres, Third edition, Paris, 1830. Vol. 2, p. 65.
2. D. H. Lehmer, On the exact number of primes less than a given limit, Illinois J. Math. 3 (1959), pp. 381–388.
3. J. C. Lagarias, V. S. Miller, and A. M. Odlyzko, Computing pi(x): The Meissel-Lehmer method, Mathematics of Computation, 44 (1985), pp. 537–560.
4. M. Deleglise and J. Rivat, "Computing pi(x): The Meissel, Lehmer, Lagarias, Miller, Odlyzko Method", Mathematics of Computation, Volume 65, Number 213, 1996, pp 235–245.
5. Hans Riesel, Prime Numbers and Computer Methods for Factorization, 2nd ed., Birkhäuser, Boston, 1994. pp. 10-38.
6. Raymond Séroul, Programming for Mathematicians, Springer-Verlag, Berlin (2000), pp. 175-181.
7. R. Crandall and C. Pomerance, Prime numbers: a computational perspective, 2nd ed., Springer, New York, 2005. pp. 152-162.
8. Tomás Oliveira e Silva, Computing pi(x): the combinatorial method, Revista do DETUA, vol. 4, no. 6, March 2006, pp. 759-768.
9. Douglas B. Staple, The combinatorial algorithm for computing pi(x), Master of Science Thesis, Dalhousie University Halifax, Nova Scotia, August 2015.


ChangeLog
2016-12-03 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.5 released.

  This version runs up to 5% faster due to a new AVX2 popcount
  algorithm which speeds up the computation of the hard special
  leaves in the range [10^17, 10^23].

  * libpopcnt.h: AVX2 & POPCNT popcount algorithm.
  * configure.ac: New --disable-popcnt, --disable-avx2.

2016-08-06 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.4 released.

  * Makfile.msvc: Use libdivide also with MSVC compiler.
  * S2LoadBalancer.cpp: Improved load balancing.
  * P2.cpp: Improved load balancing.
  * P2_mpi.cpp: Improved load balancing.
  * .travis.yml: Add static C++ code analysis using cppcheck.

2016-07-16 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.3 released.

  * README.md: Fix error in "C++ API" section.
  * S2_hard.cpp: Refactoring.
  * S2_hard_mpi.cpp: Refactoring.
  * P2.cpp: Refactoring.
  * P2_mpi.cpp: Refactoring.

2016-05-09 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.2 released.

  This version runs up to 5% faster due to an improved P2(x, a)
  implementation.

  * P2.cpp: 30% faster.
  * P2_mpi.cpp: 30% faster.
  * libdivide.h: Update to latest version.
  * autogen.sh: Print error msg if Autotools is not installed.

2016-04-02 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.1 released.

  This version runs up to 20% faster! The speed up is due to
  the usage of libdivide in S2_easy, libdivide allows to replace
  expensive integer divides with comparatively cheap
  multiplication and bitshifts.

  * S2_easy_libdivide.cpp: 40% speed up due to libdivide.
  * S2_easy_mpi_libdivide.cpp: 40% speed up due to libdivide.
  * BitSieve.cpp: Fix broken Big-Endian CPU support.

2016-03-12 Kim Walisch  <kim.walisch@gmail.com>

  Version 3.0 released.

  In this release I have merged the MPI (Message Passing Interface)
  branch into the master branch. primecount can now distribute
  computations onto cluster nodes if MPI is enabled in the build
  process (--enable-mpi).

  * doc/primecount-MPI.md: primecount MPI documentation.
  * src/mpi: primecount MPI source code.
  * include/FactorTable.hpp: Multi-threaded initialization.
  * build.sh: Improved build script with support for primecount MPI.
  * README.md: Updated build instructions.

2016-02-14 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.6 released.

  primecount has been distributed using MPI (Message Passing Interface)
  and can now run computations on large clusters!
  The distributed version of primecount is named primecount-mpi and
  the code can be found at:

  https://github.com/kimwalisch/primecount/tree/mpi

  * src/phi.cpp: 2x speed up due to pi(x) lookup table optimization.
  * scripts/build.sh: Easily build primecount.

2016-01-24 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.5 released.

  This version adds logging to primecount-backup and fixes 2
  integer overflow bugs in primecount-backup.

  * --log: Logs results into results.txt and primecount.log.
  * scripts/worktodo.sh: Script for batch processing via worktodo.txt.
  * src/S1.cpp: Fix integer overflow bug (in backup branch).
  * src/deleglise-rivat/S2_trivial.cpp: Fix integer overflow bug (in backup branch).

2015-12-27 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.4 released.

  * README.md: Simplified build instructions.
  * appveyor.yml: Automated Windows (MSVC++) testing.
  * configure.ac: Removed usage of buggy ax_gcc_builtin.m4 macro.
  * src/P2.cpp: Port to primesieve-5.5.0 library.
  * src/test.cpp: Faster nth prime testing.

2015-10-07 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.3 released.

  This version runs about 10% faster for x <= 10^21. Because of the
  sieving optimizations implemented in primecount-2.2 the sieving
  limit has now been increased which reduces the time to compute the
  easy special leaves.

  I have now also officially published primecount-backup binaries
  which save intermediate results to a file once per hour and which
  can resume from these files after e.g. a crash:
  https://github.com/kimwalisch/primecount/tree/backup

  * Renamed to --P2, --S1, --S2_trivial, --S2_easy, --S2_hard.
  * find_fastest_alpha.sh: Benchmark which finds fastest alpha factors.
  * src/primecount.cpp: Alpha factor tuning.
  * src/P2.cpp: Use multi-threading for initialization.
  * src/S2Status.cpp: --status[=N], N digits after decimal point.
  * src/pi_lehmer.cpp: Remove pi_lehmer2(x).

2015-09-19 Kim Walisch  <kim.walisch@gmail.com>

  This version runs about 10% faster due to newly added pre-sieving
  of small primes and wheel factorization.

  Version 2.2 released.

  * src/primecount.cpp: Increase pi(x) limit to 10^31 (previously 10^27).
  * src/BitSieve.cpp: Add pre-sieving of small primes.
  * src/P2.cpp: Use pre-sieving and wheel factorization.
  * src/deleglise-rivat/*: Use pre-sieving and wheel factorization.
  * src/lmo/*: Use pre-sieving and wheel factorization.
  * include/popcount.hpp: Faster popcount algorithm for CPUs without POPCNT.
  * include/Wheel.hpp: New wheel factorization data structures.

2015-04-12 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.1 released.

  * src/S1.cpp: Fix Windows performance regression.
  * src/S2LoadBalancer.cpp: Refactoring.
  * Makefile.am: Add autogen.sh to EXTRA_DIST.

2015-03-26 Kim Walisch  <kim.walisch@gmail.com>

  Version 2.0 released.

  This version improves the POPCNT algorithm in the computation of the
  hard special leaves and contains a new algorithm for the computation
  of the ordinary leaves which uses half as much memory.

  * src/S1.cpp: New implementation based on Douglas Staple's algorithm.
  * src/S2LoadBalancer.cpp: Improved load balancing.
  * src/deleglise-rivat/S2_hard.cpp: Also use POPCNT if high < y.
  * src/lmo/pi_lmo5.cpp: Optimized sieving, up to 10% faster.
  * src/lmo/pi_lmo_parallel3.cpp: Optimized sieving, up to 10% faster.
  * include/BitSieve.hpp: Add count backwards optimization.

2015-03-08 Kim Walisch  <kim.walisch@gmail.com>

  Version 1.9 released.


include/PiTable.hpp
///
/// @file   PiTable.hpp
/// @brief  The PiTable class is a compressed lookup table for prime
///         counts. It uses only n / 64 * 16 bytes of memory and
///         returns the number of primes below n using O(1) operations.
///
///         How it works:
///
///         1) pi_[n / 64].prime_count = pi(n - n % 64).
///
///         2) pi_[n / 64].bits corresponds to ]base, base + 64] with
///            base = n - n % 64, each bit represents one number and
///            if the bit is 1 the number is a prime. Using the
///            POPCNT instruction we can now easily count the primes
///            within ]n - n % 64, n] and add it to 1).
///
/// Copyright (C) 2016 Kim Walisch, <kim.walisch@gmail.com>
///
/// This file is distributed under the BSD License. See the COPYING
/// file in the top level directory.
///

#ifndef PITABLE_HPP
#define PITABLE_HPP

#if !defined(__STDC_CONSTANT_MACROS)
  #define __STDC_CONSTANT_MACROS
#endif

#include <libpopcnt.h>

#include <stdint.h>
#include <cassert>
#include <vector>

namespace primecount {

class PiTable
{
public:
  PiTable(uint64_t max);

  /// @return  The number of primes <= n
  int64_t operator[](uint64_t n) const
  {
    assert(n <= max_);
    uint64_t bitmask = UINT64_C(0xffffffffffffffff) >> (63 - n % 64);
    return pi_[n / 64].prime_count + popcnt_u64(pi_[n / 64].bits & bitmask);
  }

  int64_t size() const
  {
    return max_ + 1;
  }
private:
  struct PiData
  {
    PiData() : prime_count(0), bits(0) { }
    uint64_t prime_count;
    uint64_t bits;
  };

  std::vector<PiData> pi_;
  uint64_t max_;
};

} // namespace

#endif


scripts/update_version.sh
#!/bin/sh

if [ $# -ne 2 ]
then
    echo "Usage example:"
    echo "$ ./update_version.sh 1.2.3 7:0:0"
    echo ""
    echo "Updates the primecount version to 1.2.3 in all files and"
    echo "updates the libtool library version to 7:0:0"

    exit 1
fi

# Run from primecount root directory
test -e ../src && cd ..

new_version=$1
old_version=$(grep "PRIMECOUNT_VERSION " include/primecount.hpp | cut -f2 -d'"')

new_major=$(echo $new_version | cut -f1 -d'.')
new_minor=$(echo $new_version | cut -f2 -d'.')

old_major=$(echo $old_version | cut -f1 -d'.')
old_minor=$(echo $old_version | cut -f2 -d'.')

new_libtool_version=$2
old_libtool_version=$(grep primecount_lib_version configure.ac | \
                      cut -f2 -d' ' | \
                      cut -f2 -d'[' | \
                      cut -f1 -d']')

new_year=$(date +'%Y')
old_year=$(grep "Copyright (c)" COPYING | cut -f5 -d' ' | cut -f1 -d',')

echo "New version: $new_version"
echo "Old version: $old_version"
echo ""
echo "New libtool version: $new_libtool_version"
echo "Old libtool version: $old_libtool_version"
echo ""
echo "New year: $new_year"
echo "Old year: $old_year"
echo ""

# Update version
for i in $(echo README.md \
                configure.ac \
                doc/primecount-MPI.md \
                include/primecount.hpp)
do
    echo "Update version in $i"
    sed "s/$old_major\.$old_minor/$new_version/g" $i > $i.tmp
    mv -f $i.tmp $i
done

# Update version
for i in $(echo include/primecount.hpp)
do
    sed "s/PRIMECOUNT_VERSION_MAJOR $old_major/PRIMECOUNT_VERSION_MAJOR $new_major/g" $i > $i.tmp
    mv -f $i.tmp $i
    sed "s/PRIMECOUNT_VERSION_MINOR $old_minor/PRIMECOUNT_VERSION_MINOR $new_minor/g" $i > $i.tmp
    mv -f $i.tmp $i
done

# Update libtool version
echo ""
echo "Update libtool version in configure.ac $new_libtool_version"
sed "s/$old_libtool_version/$new_libtool_version/g" configure.ac > configure.ac.tmp
mv -f configure.ac.tmp configure.ac
echo ""

# Update year
for i in $(echo COPYING \
                src/app/help.cpp)
do
    echo "Update year in $i"
    sed "s/$old_year/$new_year/g" $i > $i.tmp
    mv -f $i.tmp $i
done

echo "Version has been updated!"


src/BitSieve.cpp
///
/// @file  BitSieve.cpp
/// @brief The BitSieve class is a bit array for use with
///        Eratosthenes-like prime sieving algorithms. BitSieve
///        assigns 64 numbers to the bits of an 8 byte word thus
///        reducing the memory usage by a factor of 8.
///
/// Copyright (C) 2016 Kim Walisch, <kim.walisch@gmail.com>
///
/// This file is distributed under the BSD License. See the COPYING
/// file in the top level directory.
///

#if !defined(__STDC_CONSTANT_MACROS)
  #define __STDC_CONSTANT_MACROS
#endif

#include <BitSieve.hpp>
#include <pmath.hpp>
#include <libpopcnt.h>

#include <stdint.h>
#include <algorithm>
#include <cassert>
#include <cstring>
#include <vector>

using namespace std;

namespace {

/// 1-indexing: primes[1] = 2, primes[2] = 3, ...
const uint64_t primes[] = { 0, 2, 3, 5, 7, 11, 13, 17, 19, 23 };

/// Bitmasks with multiples of the i-th prime set
const uint64_t masks[] =
{
  UINT64_C(0x0000000000000000),
  UINT64_C(0x5555555555555555), // 2
  UINT64_C(0x9249249249249249), // 3
  UINT64_C(0x1084210842108421), // 5
  UINT64_C(0x8102040810204081), // 7
  UINT64_C(0x0080100200400801), // 11
  UINT64_C(0x0010008004002001), // 13
  UINT64_C(0x0008000400020001), // 17
  UINT64_C(0x0200004000080001), // 19
  UINT64_C(0x0000400000800001)  // 23
};

/// Get bitmask with unset multiples
uint64_t unset_mask(uint64_t mask, uint64_t shift)
{
  return ~(mask << shift);
}

/// @pre x < y * 2
uint64_t fast_modulo(uint64_t x, uint64_t y)
{
  x = (x < y) ? x : x - y;
  assert(x < y);
  return x;
}

}

namespace primecount {

const uint64_t BitSieve::unset_bit_[64] =
{
  ~(UINT64_C(1) <<  0), ~(UINT64_C(1) <<  1), ~(UINT64_C(1) <<  2),
  ~(UINT64_C(1) <<  3), ~(UINT64_C(1) <<  4), ~(UINT64_C(1) <<  5),
  ~(UINT64_C(1) <<  6), ~(UINT64_C(1) <<  7), ~(UINT64_C(1) <<  8),
  ~(UINT64_C(1) <<  9), ~(UINT64_C(1) << 10), ~(UINT64_C(1) << 11),
  ~(UINT64_C(1) << 12), ~(UINT64_C(1) << 13), ~(UINT64_C(1) << 14),
  ~(UINT64_C(1) << 15), ~(UINT64_C(1) << 16), ~(UINT64_C(1) << 17),
  ~(UINT64_C(1) << 18), ~(UINT64_C(1) << 19), ~(UINT64_C(1) << 20),
  ~(UINT64_C(1) << 21), ~(UINT64_C(1) << 22), ~(UINT64_C(1) << 23),
  ~(UINT64_C(1) << 24), ~(UINT64_C(1) << 25), ~(UINT64_C(1) << 26),
  ~(UINT64_C(1) << 27), ~(UINT64_C(1) << 28), ~(UINT64_C(1) << 29),
  ~(UINT64_C(1) << 30), ~(UINT64_C(1) << 31), ~(UINT64_C(1) << 32),
  ~(UINT64_C(1) << 33), ~(UINT64_C(1) << 34), ~(UINT64_C(1) << 35),
  ~(UINT64_C(1) << 36), ~(UINT64_C(1) << 37), ~(UINT64_C(1) << 38),
  ~(UINT64_C(1) << 39), ~(UINT64_C(1) << 40), ~(UINT64_C(1) << 41),
  ~(UINT64_C(1) << 42), ~(UINT64_C(1) << 43), ~(UINT64_C(1) << 44),
  ~(UINT64_C(1) << 45), ~(UINT64_C(1) << 46), ~(UINT64_C(1) << 47),
  ~(UINT64_C(1) << 48), ~(UINT64_C(1) << 49), ~(UINT64_C(1) << 50),
  ~(UINT64_C(1) << 51), ~(UINT64_C(1) << 52), ~(UINT64_C(1) << 53),
  ~(UINT64_C(1) << 54), ~(UINT64_C(1) << 55), ~(UINT64_C(1) << 56),
  ~(UINT64_C(1) << 57), ~(UINT64_C(1) << 58), ~(UINT64_C(1) << 59),
  ~(UINT64_C(1) << 60), ~(UINT64_C(1) << 61), ~(UINT64_C(1) << 62),
  ~(UINT64_C(1) << 63)
};

BitSieve::BitSieve(std::size_t size) :
  sieve_(ceil_div(size, 64)),
  size_(size)
{ }

/// Pre-sieve the multiples (>= low) of the first c primes.
/// @warning Also removes the first c primes.
/// @pre c < 10
///
void BitSieve::pre_sieve(uint64_t c, uint64_t low)
{
  assert(c < 10);

  if (sieve_.empty())
    return;

  // unset multiples of 2
  sieve_[0] = unset_mask(masks[1], low % 2);

  uint64_t last = 1;
  uint64_t sieved = last;
  uint64_t sieve_size = sieve_.size();

  // pre-sieve multiples of first c primes
  for (uint64_t i = 2; i <= c; i++)
  {
    uint64_t prime = primes[i];
    uint64_t end_copy = min(sieved * prime, sieve_size);

    // pre-sieve multiples of primes < i-th prime
    // by copying a small pre-sieved buffer
    while (last < end_copy)
    {
      uint64_t copy_words = min(sieved, sieve_size - last);
      copy(sieve_.begin(),
           sieve_.begin() + copy_words,
           sieve_.begin() + last);
      last += copy_words;
    }

    // calculate first multiple >= low of prime
    uint64_t multiple = ceil_div(low, prime) * prime;
    uint64_t shift = multiple - low;
    uint64_t next_shift = prime - 64 % prime;
    uint64_t mask = masks[i];

    // pre-sieve multiples of the i-th prime
    for (uint64_t j = 0; j < last; j++)
    {
      sieve_[j] &= unset_mask(mask, shift);
      shift = fast_modulo(shift + next_shift, prime);
    }

    sieved = last;
  }

  // fill up the rest of the sieve
  while (last < sieve_size)
  {
    uint64_t copy_words = min(sieved, sieve_size - last);
    copy(sieve_.begin(),
         sieve_.begin() + copy_words,
         sieve_.begin() + last);
    last += copy_words;
  }
}

/// Count the number of 1 bits inside [start, stop]
uint64_t BitSieve::count(uint64_t start,
                         uint64_t stop) const
{
  if (start > stop)
    return 0;

  assert(stop < size_);

  uint64_t start_idx = start / 64;
  uint64_t stop_idx = stop / 64;
  uint64_t m1 = UINT64_C(0xffffffffffffffff) << (start % 64);
  uint64_t m2 = UINT64_C(0xffffffffffffffff) >> (63 - stop % 64);
  uint64_t bit_count;

  if (start_idx == stop_idx)
    bit_count = popcnt_u64(sieve_[start_idx] & (m1 & m2));
  else
  {
    bit_count = popcnt_u64(sieve_[start_idx] & m1);
    bit_count += popcnt(&sieve_[start_idx + 1], (stop_idx - start_idx - 1) * sizeof(uint64_t));
    bit_count += popcnt_u64(sieve_[stop_idx] & m2);
  }

  return bit_count;
}

} // namespace
